client/src/main/java/io/getstream/chat/android/client/utils/observable/Subscriptions.kt

```diff
@@ -0,0 +1,52 @@
+package io.getstream.chat.android.client.utils.observable
+
+import io.getstream.chat.android.client.events.ChatEvent
+
+interface Disposable {
+    val isDisposed: Boolean
+    fun dispose()
+}
+
+internal interface EventSubscription : Disposable {
+    fun onNext(event: ChatEvent)
+}
+
+internal open class SubscriptionImpl(
+    private val filter: (ChatEvent) -> Boolean,
+    listener: ((ChatEvent) -> Unit)
+) : EventSubscription {
+
+    private var listener: ((ChatEvent) -> Unit)? = listener
+
+    override var isDisposed: Boolean = false
+
+    override fun dispose() {
+        isDisposed = true
+        listener = null
+    }
+
+    final override fun onNext(event: ChatEvent) {
+        check(!isDisposed) { "Subscription already disposed, onNext should not be called on it" }
+
+        if (filter(event)) {
+            try {
+                listener!!.invoke(event)
+            } finally {
+                afterEventDelivered()
+            }
+        }
+    }
+
+    open fun afterEventDelivered() {}
```

Comment:
We could add this listener as a setter in this class and make it a final class. In this case I would prefer go with composition instead of inheritance.